File: src/main/java/wraith/fabricaeexnihilo/modules/sieves/SieveBlockEntity.java
package wraith.fabricaeexnihilo.modules.sieves;

import net.fabricmc.fabric.api.object.builder.v1.block.entity.FabricBlockEntityTypeBuilder;
import net.minecraft.block.BlockState;
import net.minecraft.block.entity.BlockEntityType;
import net.minecraft.enchantment.EnchantmentHelper;
import net.minecraft.enchantment.Enchantments;
import net.minecraft.entity.effect.StatusEffects;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.fluid.Fluid;
import net.minecraft.fluid.Fluids;
import net.minecraft.item.BucketItem;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.registry.Registries;
import net.minecraft.state.property.Properties;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Hand;
import net.minecraft.util.Identifier;
import net.minecraft.util.ItemScatterer;
import net.minecraft.util.hit.BlockHitResult;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.world.World;
import org.jetbrains.annotations.Nullable;
import wraith.fabricaeexnihilo.FabricaeExNihilo;
import wraith.fabricaeexnihilo.modules.ModBlocks;
import wraith.fabricaeexnihilo.modules.base.BaseBlockEntity;
import wraith.fabricaeexnihilo.recipe.ModRecipes;
import wraith.fabricaeexnihilo.recipe.SieveRecipe;
import wraith.fabricaeexnihilo.util.ItemUtils;

import java.util.*;

import static wraith.fabricaeexnihilo.FabricaeExNihilo.id;
import static wraith.fabricaeexnihilo.modules.sieves.SieveBlock.WATERLOGGED;

public class SieveBlockEntity extends BaseBlockEntity {

    public static final BlockEntityType<SieveBlockEntity> TYPE = FabricBlockEntityTypeBuilder.create(
            SieveBlockEntity::new,
            ModBlocks.SIEVES.values().toArray(new SieveBlock[0])
    ).build(null);
    public static final Identifier BLOCK_ENTITY_ID = id("sieve");
    double progress = 0.0;
    private ItemStack contents = ItemStack.EMPTY;
    private ItemStack mesh = ItemStack.EMPTY;

    public SieveBlockEntity(BlockPos pos, BlockState state) {
        super(TYPE, pos, state);
    }

    public ActionResult activate(BlockState state, PlayerEntity player, Hand hand, BlockHitResult hitResult) {
        if (world == null || player == null) {
            return ActionResult.PASS;
        }

        var held = player.getStackInHand(hand == null ? player.getActiveHand() : hand);
        if (held == null) {
            held = ItemStack.EMPTY;
        }

        if (held.getItem() instanceof BucketItem) {
            return ActionResult.PASS; // Done for fluid logging
        }

        var sieves = getConnectedSieves();
        // Make Progress
        if (!contents.isEmpty()) {
            sieves.forEach(sieve -> sieve.doProgress(player));
            return ActionResult.SUCCESS;
        }

        var item = held.getItem();
        // Removing mesh
        if (!mesh.isEmpty() && player.getMainHandStack().isEmpty()) {
            player.getInventory().offerOrDrop(mesh.copy());
            mesh = ItemStack.EMPTY;
            markDirty();
            return ActionResult.SUCCESS;
        } else if (mesh.isEmpty() &&
                world.getRecipeManager()
                        .listAllOfType(ModRecipes.SIEVE)
                        .stream()
                        .anyMatch(recipe -> recipe.getRolls().containsKey(Registries.ITEM.getId(item)))
        ) {
            // Add mesh
            mesh = ItemUtils.ofSize(held, 1);
            if (!player.isCreative()) {
                held.decrement(1);
            }
            markDirty();
            return ActionResult.SUCCESS;
        }

        // Add a block
        if (held.isEmpty() || SieveRecipe.find(held.getItem(), state.get(WATERLOGGED), Registries.ITEM.getId(mesh.getItem()), world).isEmpty()) {
            return ActionResult.PASS;
        }
        final ItemStack finalHeld = held;
        sieves.forEach(sieve -> sieve.setContents(finalHeld, !player.isCreative()));
        return ActionResult.SUCCESS;
    }

    public void doProgress(PlayerEntity player) {
        if (world == null) {
            return;
        }
        var haste = player.getActiveStatusEffects().get(StatusEffects.HASTE);
        var efficiency = FabricaeExNihilo.CONFIG.modules.sieves.efficiency ? EnchantmentHelper.getLevel(Enchantments.EFFICIENCY, mesh) : 0;
        var hasteLevel = FabricaeExNihilo.CONFIG.modules.sieves.haste ? (haste == null ? -1 : haste.getAmplifier()) + 1 : 0;

        progress += FabricaeExNihilo.CONFIG.modules.sieves.baseProgress
                + FabricaeExNihilo.CONFIG.modules.sieves.efficiencyScaleFactor * efficiency
                + FabricaeExNihilo.CONFIG.modules.sieves.hasteScaleFactor * hasteLevel;

        //TODO: spawn some particles
        if (progress > 1.0) {
            // The utility method for multiple items is less neat to use
            for (var result : SieveRecipe.find(contents.getItem(), getCachedState().get(WATERLOGGED), Registries.ITEM.getId(mesh.getItem()), world)) {
                ItemScatterer.spawn(world, pos.getX(), pos.getY() + 1, pos.getZ(), result.createStack(world.random));
            }
            progress = 0.0;
            contents = ItemStack.EMPTY;
        }
        markDirty();
    }

    public void dropInventory() {
        ItemScatterer.spawn(world, pos.getX(), pos.getY() + 1, pos.getZ(), mesh);
        ItemScatterer.spawn(world, pos.getX(), pos.getY() + 1, pos.getZ(), contents);
        mesh = ItemStack.EMPTY;
        contents = ItemStack.EMPTY;
    }

    public List<SieveBlockEntity> getConnectedSieves() {
        var sieves = new ArrayList<SieveBlockEntity>();

        if (world == null) {
            return sieves;
        }

        var tested = new HashSet<BlockPos>();
        var stack = new Stack<BlockPos>();
        stack.add(this.pos);

        while (!stack.empty()) {
            var popped = stack.pop();
            // Record that this one has been tested
            tested.add(popped);
            if (matchingSieveAt(world, popped)) {
                if (!(this.world.getBlockEntity(popped) instanceof SieveBlockEntity sieve)) {
                    continue;
                }
                sieves.add(sieve);
                // Add adjacent locations to test to the stack
                Arrays.stream(Direction.values())
                        // Horizontals
                        .filter(dir -> dir.getAxis().isHorizontal())
                        // to BlockPos
                        .map(popped::offset)
                        // Remove already tested positions
                        .filter(dir -> !tested.contains(dir) && !stack.contains(dir))
                        // Remove positions too far away
                        .filter(dir -> Math.abs(this.pos.getX() - dir.getX()) <= FabricaeExNihilo.CONFIG.modules.sieves.sieveRadius &&
                                Math.abs(this.pos.getZ() - dir.getZ()) <= FabricaeExNihilo.CONFIG.modules.sieves.sieveRadius)
                        // Add to the stack to be processed
                        .forEach(stack::add);
            }
        }

        return sieves;
    }

    public ItemStack getContents() {
        return contents;
    }

    @Nullable
    public Fluid getFluid() {
        if (world == null) {
            return null;
        }
        var state = world.getBlockState(pos);
        if (state == null) {
            return null;
        }
        if (!state.get(Properties.WATERLOGGED)) {
            return Fluids.EMPTY;
        }
        return state.getFluidState().getFluid();
    }

    public ItemStack getMesh() {
        return mesh;
    }

    public double getProgress() {
        return progress;
    }

    private boolean matchingSieveAt(@Nullable World world, BlockPos pos) {
        if (world == null) {
            return false;
        }
        if (world.getBlockEntity(pos) instanceof SieveBlockEntity sieve) {
            return ItemStack.areItemsEqual(mesh, sieve.mesh) && world.getFluidState(pos).getFluid() == getFluid();
        }
        return false;
    }

    @Override
    public void readNbt(NbtCompound nbt) {
        super.readNbt(nbt);
        if (nbt == null) {
            FabricaeExNihilo.LOGGER.warn("A sieve at " + this.pos + " is missing data.");
            return;
        }
        readNbtWithoutWorldInfo(nbt);
    }

    public void readNbtWithoutWorldInfo(NbtCompound nbt) {
        mesh = ItemStack.fromNbt(nbt.getCompound("mesh"));
        contents = ItemStack.fromNbt(nbt.getCompound("contents"));
        progress = nbt.getDouble("progress");
    }

    public void setContents(ItemStack stack, boolean doDrain) {
        if (stack.isEmpty() || !contents.isEmpty()) {
            return;
        }
        contents = ItemUtils.ofSize(stack, 1);
        if (doDrain) {
            stack.decrement(1);
        }
        progress = 0.0;
        markDirty();
    }

    /**
     * NBT Serialization section
     */

    @Override
    public void writeNbt(NbtCompound nbt) {
        super.writeNbt(nbt);
        writeNbtWithoutWorldInfo(nbt);
    }

    public void writeNbtWithoutWorldInfo(NbtCompound nbt) {
        nbt.put("mesh", mesh.writeNbt(new NbtCompound()));
        nbt.put("contents", contents.writeNbt(new NbtCompound()));
        nbt.putDouble("progress", progress);
    }


}

File: src/main/kotlin/exnihilofabrico/modules/sieves/SieveBlockEntity.kt
package exnihilofabrico.modules.sieves

import exnihilofabrico.ExNihiloFabrico
import exnihilofabrico.api.registry.ExNihiloRegistries.SIEVE
import exnihilofabrico.id
import exnihilofabrico.modules.ModBlocks
import exnihilofabrico.modules.base.BaseBlockEntity
import exnihilofabrico.util.ofSize
import net.fabricmc.fabric.api.block.entity.BlockEntityClientSerializable
import net.minecraft.block.BlockState
import net.minecraft.block.entity.BlockEntityType
import net.minecraft.enchantment.EnchantmentHelper
import net.minecraft.enchantment.Enchantments
import net.minecraft.entity.effect.StatusEffects
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.fluid.Fluid
import net.minecraft.fluid.Fluids
import net.minecraft.item.BucketItem
import net.minecraft.item.ItemStack
import net.minecraft.nbt.CompoundTag
import net.minecraft.util.DefaultedList
import net.minecraft.util.Hand
import net.minecraft.util.ItemScatterer
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.World
import virtuoel.towelette.api.Fluidloggable
import java.util.*
import kotlin.collections.HashSet
import kotlin.math.abs

class SieveBlockEntity: BaseBlockEntity(TYPE), BlockEntityClientSerializable {

    var mesh: ItemStack = ItemStack.EMPTY
    var contents: ItemStack = ItemStack.EMPTY
    var progress: Double = 0.0

    fun activate(state: BlockState?, player: PlayerEntity?, hand: Hand?, hitResult: BlockHitResult?): Boolean {

        if(world?.isClient != false || player == null)
            return true

        val held = player.getStackInHand(hand ?: player.activeHand) ?: ItemStack.EMPTY

        if(held.item is BucketItem)
            return false // Done for fluid logging


        // Remove/Swap a mesh
        if(SIEVE.isValidMesh(held)) {
            // Removing  mesh
            if(!mesh.isEmpty) {
                player.giveItemStack(mesh.copy())
                mesh = ItemStack.EMPTY
            }
            // Add mesh
            if(SIEVE.isValidMesh(held)) {
                mesh = held.ofSize(1)
                if(!player.isCreative)
                    held.decrement(1)
            }
            markDirtyClient()
            return true
        }
        val sieves = getConnectedSieves()
        // Make Progress
        if(!contents.isEmpty) {
            sieves.forEach { it.doProgress(player) }
            return true
        }

        // Add a block
        if(!held.isEmpty && SIEVE.isValidRecipe(mesh, getFluid(), held)) {
            sieves.forEach { it.setContents(held, !player.isCreative) }
            return true
        }
        return false
    }

    fun setContents(stack: ItemStack, doDrain: Boolean) {
        if(stack.isEmpty || !contents.isEmpty) return
        contents = stack.ofSize(1)
        if(doDrain)
            stack.decrement(1)
        progress = 0.0
        markDirtyClient()
    }

    fun getFluid(): Fluid? {
        val state = world?.getBlockState(pos) ?: return null
        if(state.block !is Fluidloggable) return Fluids.EMPTY
        return state.fluidState.fluid
    }

    fun doProgress(player: PlayerEntity) {

        val efficiency = if(ExNihiloFabrico.config.modules.sieves.efficiency)
            EnchantmentHelper.getLevel(Enchantments.EFFICIENCY, mesh)
        else
            0
        val haste = if(ExNihiloFabrico.config.modules.sieves.haste)
            (player.activeStatusEffects[StatusEffects.HASTE]?.amplifier ?: -1) + 1
        else
            0

        progress += ExNihiloFabrico.config.modules.sieves.baseProgress
                + ExNihiloFabrico.config.modules.sieves.efficiencyScaleFactor * efficiency
                + ExNihiloFabrico.config.modules.sieves.hasteScaleFactor * haste

        // TODO spawn some particles

        if(progress > 1.0) {
            SIEVE.getResult(mesh, getFluid(), contents, player, world?.random ?: return)
                .forEach {player.giveItemStack(it)}
            progress = 0.0
            contents = ItemStack.EMPTY
        }
        markDirtyClient()
    }

    fun dropInventory() {
        ItemScatterer.spawn(world, pos.up(), DefaultedList.copyOf(mesh, contents))
        mesh = ItemStack.EMPTY
        contents = ItemStack.EMPTY
    }

    fun dropMesh() {
        ItemScatterer.spawn(world, pos.up(), DefaultedList.copyOf(mesh))
        mesh = ItemStack.EMPTY
    }


    fun dropContents() {
        ItemScatterer.spawn(world, pos.up(), DefaultedList.copyOf(contents))
        contents = ItemStack.EMPTY
    }

    fun getConnectedSieves(): Collection<SieveBlockEntity> {
        val sieves = mutableListOf<SieveBlockEntity>()

        val tested = HashSet<BlockPos>()
        val stack = Stack<BlockPos>()
        stack.add(this.pos)
        while(!stack.empty()) {
            val popped = stack.pop()
            // Record that this one has been tested
            tested.add(popped)
            if(matchingSieveAt(world?: continue, popped)) {
                sieves.add(this.world?.getBlockEntity(popped) as? SieveBlockEntity ?: continue)
                // Add adjacent locations to test to the stack
                Direction.values()
                    // Horizontals
                    .filter{
                        it.offsetY == 0 }
                    // to BlockPos
                    .map { popped.offset(it) }
                    // Remove already tested positions
                    .filter { !tested.contains(it) && !stack.contains(it) }
                    // Remove positions too far away
                    .filter {
                        abs(this.pos.x - it.x) <= ExNihiloFabrico.config.modules.sieves.sieveRadius &&
                            abs(this.pos.z - it.z) <= ExNihiloFabrico.config.modules.sieves.sieveRadius }
                    // Add to the stack to be processed
                    .forEach { stack.add(it) }
            }
        }

        return sieves
    }

    private fun matchingSieveAt(world: World, pos: BlockPos): Boolean {
        val other = world.getBlockEntity(pos) as? SieveBlockEntity ?: return false
        return ItemStack.areItemsEqual(mesh, other.mesh)
    }

    /**
     * NBT Serialization section
     */

    override fun toTag(tag: CompoundTag) = toTagWithoutWorldInfo(super.toTag(tag))

    override fun fromTag(tag: CompoundTag?) {
        super.fromTag(tag)
        if(tag==null){
            ExNihiloFabrico.LOGGER.warn("A sieve at $pos is missing data.")
            return
        }
        fromTagWithoutWorldInfo(tag)
    }

    override fun toClientTag(tag: CompoundTag?) = toTag(tag ?: CompoundTag())
    override fun fromClientTag(tag: CompoundTag?) = fromTag(tag ?: CompoundTag())

    fun toTagWithoutWorldInfo(tag: CompoundTag): CompoundTag {
        tag.put("mesh", mesh.toTag(CompoundTag()))
        tag.put("contents", contents.toTag(CompoundTag()))
        tag.putDouble("progress", progress)
        return tag
    }

    fun fromTagWithoutWorldInfo(tag: CompoundTag) {
        mesh = ItemStack.fromTag(tag.getCompound("mesh"))
        contents = ItemStack.fromTag(tag.getCompound("contents"))
        progress = tag.getDouble("progress")
    }

    companion object {
        @JvmStatic
        val TYPE: BlockEntityType<SieveBlockEntity> =
            BlockEntityType.Builder.create({SieveBlockEntity()}, ModBlocks.SIEVES.values.toTypedArray()).build(null)
        val BLOCK_ENTITY_ID = id("sieve")
    }
}